CMakeLists: Generate hierarchical source groups for IDEs

fwcd opened the pull request:
This is a largely cosmetic improvement, but when generating a Visual Studio or Xcode project from Mixxx, the default is to generate a flat list of source files. For editing, however, it would be much nicer to have the same directory structure as we do in the file system. Therefore this PR uses the [`source_group(TREE ...)`](https://cmake.org/cmake/help/latest/command/source_group.html) command[^1] to mimic this layout.

For comparison, this is how it looks in Xcode:

| Before | After |
| - | - |
| <img src="https://github.com/user-attachments/assets/e8bf8c8e-3ef3-4a41-a242-08c1e0eac79b" height="500"> | <img src="https://github.com/user-attachments/assets/174acef3-f246-409f-8b09-ac93360a44d2" height="500"> |


[^1]: As e.g. suggested in https://stackoverflow.com/a/56496104

fwcd committed:
CMakeLists: Generate hierarchical source groups for IDEs

daschuer reviewed (commented):
Sounds good. Do we want to apply it to 2.4 to make it easier to change between branches (one day)?

JoergAtGithub commented:
In my VisualStudio it appears already structured without this PR:
![grafik](https://github.com/user-attachments/assets/cc2504ea-740f-455c-8632-e43da0f391a2)

fwcd commented:
2.4 or 2.5? Targeting main was unintentional, we should probably go with 2.5?

fwcd commented:
@JoergAtGithub Interesting, that's the precompiled header group though. Could you check the contents of "Source Files"?

JoergAtGithub commented:
> hat's the precompiled header group though

I don't think so, precompiled header is collapsed. Here it's expanded pch tree:
![grafik](https://github.com/user-attachments/assets/09ec101c-3986-4304-ac01-2006addbab00)

fwcd commented:
Oh right, I misread, you're right

fwcd commented:
Just out of curiosity, does CMake generate an additional "Source Files" group in your VS project?

JoergAtGithub commented:
I'm not sure, what you mean, but these are all targets listed after CMake configure completes:
![grafik](https://github.com/user-attachments/assets/67a723da-28d5-48b9-8f1b-194bea1cc0f1)

fwcd commented:
I mean within mixxx-lib. I think VS calls groups "filters"?

fwcd commented:
I have rebased the PR to 2.5, so it should be ready for re-review.

JoergAtGithub commented:
Now I see two subtrees, `Source Files` and `src`:
![grafik](https://github.com/user-attachments/assets/243e8221-860f-4a6a-95e7-e44a39edca76)

fwcd commented:
Interesting, so it looks like either Visual Studio or CMake are implicitly adding the `src` tree. I wonder why `coreservices.cpp` et al. show up under `src` though, in Xcode they are also placed under `Source Files`...

Maybe we should just scope this logic to the Xcode generator for now since the VS project already generates a hierarchy?

fwcd committed:
CMakeLists: Only generate source groups for Xcode

Visual Studio already seems to provide a hierarchical view of the source
tree, so this is not needed.

daschuer commented:
Do you know how it relates to visual studio. Does this also improve the situation there? Are files not added to mixxx-lib still displayed reasonable?

fwcd commented:
Haven't had the chance to test it with Visual Studio, but according to @JoergAtGithub it made the hierarchy a bit more confusing there since VS already seems to group files hierarchically. Hence why I've scoped the logic to Xcode for now, since it definitely improves the situation there.

fwcd committed:
CMakeLists: Fix source_group formatting

fwcd commented:
I've moved this to `main`, hopefully this should also address the new style checks

acolombier commented:
Any objection to merge that? This looks like a net developer experience for XCode user.

daschuer reviewed (approved):
Looks good. Can you re-base that to 2.5 to give a unified development experience on all active branches.
